### PR TITLE
(Issue #573) Update Random Number Generation to Use NumPy Random Generators

### PR DESCRIPTION
**Pull request by samadpls**
This pull request addresses issue #573 by updating the random number generation in the codebase. It replaces the deprecated use of `numpy.random.RandomState` with the recommended approach of using NumPy random generators (`numpy.random.Generator`).

### CODE REVIEW

**Review by AMHermansen — COMMENTED** [review]
Hello @samadpls :wave: thank you for this PR.

By searching for np.random in the GitHub repository I found at least one other instance where deprecated random number generation is used: `data/dataset/sqlite/sqlite_dataset_perturbed.py`. Line 130

Since this PR aims to remove the legacy random number generation from numpy to the new supported version I think we should also update how it's done in `data/dataset/sqlite/sqlite_dataset_perturbed.py`, possibly by adding a new optional seed or a `np.Generator` argument in the constructor. :smile: 
Let me know what you think

**Comment by samadpls** [comment]
> Hello @samadpls 👋 thank you for this PR.
> 
> By searching for np.random in the GitHub repository I found at least one other instance where deprecated random number generation is used: `data/dataset/sqlite/sqlite_dataset_perturbed.py`. Line 130
> 
> Since this PR aims to remove the legacy random number generation from numpy to the new supported version I think we should also update how it's done in `data/dataset/sqlite/sqlite_dataset_perturbed.py`, possibly by adding a new optional seed or a `np.Generator` argument in the constructor. 😄 Let me know what you think

Hi @AMHermansen  👋, thank you for the comment, I've added an optional seed argument in the constructor. Let me know if you have any further suggestions

**Review by AMHermansen — APPROVED** [review]
Looks good :smiley:

File: src/graphnet/data/dataset/sqlite/sqlite_dataset_perturbed.py [hunk 1]
```diff
@@ -78,6 +80,10 @@ def __init__(
                 in this case to events with no value in the corresponding
                 table/column. That is, if no per-event loss weight table/column
                 is provided, this value is ignored. Defaults to None.
+            seed: Optional seed for random number generation (int or numpy Generator).
```
Comment (AMHermansen) [inline src/graphnet/data/dataset/sqlite/sqlite_dataset_perturbed.py:83]: It seems the lines in the docstring are too long. Since we're using type hints it is not necessary to include the allowed values in the docstring.

File: src/graphnet/training/utils.py [hunk 2]
```diff
@@ -115,7 +115,7 @@ def make_train_validation_dataloader(
 ) -> Tuple[DataLoader, DataLoader]:
     """Construct train and test `DataLoader` instances."""
     # Reproducibility
-    rng = np.random.RandomState(seed=seed)
+    rng = np.random.default_rng(seed=seed)
```
Comment (AMHermansen) [inline src/graphnet/training/utils.py:118]: It looks like the "rng" is passed to `sklearn.model_selection.train_test_split` and `pandas.DataFrame.sample`. Maybe we should just pass the seed value directly to sklearn.model_selection.train_test_split?

**Review by AMHermansen — CHANGES_REQUESTED** [review]
@samadpls I stand corrected by the unittests / code styling.
I've left some comments which should hopefully make the tests pass, you could try and install pre-commit hooks as well to make sure the code quality passes :smiley:

File: src/graphnet/training/utils.py (hunk 2 above)
Comment (samadpls) [reply src/graphnet/training/utils.py:118]: Do you mean changing `random_state=rng` to `random_state=seed`? If yes, then we can also remove `rng` as it will not be used anywhere.

Comment (AMHermansen) [reply src/graphnet/training/utils.py]: I thought about passing the `np.Generator ` to `pd.DataFrame.sample` and the seed or state to `sklearn.model_selection.train_test_split`, but I don't see any issue passing the seed to both.

Comment (samadpls) [reply src/graphnet/training/utils.py:118]: I apologize for misunderstanding of your comment.

**Comment by AMHermansen** [comment]
@RasmusOrsoe Codeclimate is currently failing for this PR due to too high "Cognitive Complexity" should we ignore it here, or does it need refactoring?

**Comment by samadpls** [comment]
> @RasmusOrsoe Codeclimate is currently failing for this PR due to too high "Cognitive Complexity" should we ignore it here, or does it need refactoring?

not sure if adding `seed` in the parameter to the `super().__init__` fix it reduce the cognitive complexity of the code

**Comment by RasmusOrsoe** [comment]
Hello @samadpls - thank you for this contribution!

Automated checks are running; let's see what they say.

The codeclimate issue is not a problem @AMHermansen. We can merge if the rest is OK.

**Review by AMHermansen — APPROVED** [review]
Looks good
